tests/test_ext_autosummary.py

```diff
@@ -222,6 +224,115 @@ def test_autosummary_generate(app, status, warning):
+    module_summary = '.. autosummary::\n'\
+                     '   :toctree: {}\n'\
+                     '\n'\
+                     '   {{}}.{}\n'\
+                     '   {{}}.{}\n'.format(module_rubic_name,
+                                           module_name,
+                                           package_name)
+    last_module_summary = '.. autosummary::\n'\
+                          '   :toctree: {}\n'\
+                          '\n'\
+                          '   {{}}.{}\n'.format(module_rubic_name,
+                                                module_name)
+    class_summary = '   .. autosummary::\n'\
+                    '   \n'\
+                    '      {}\n'\
+                    '   \n'.format(class_name)
+
+    def assert_package_content(file):
+        pkgroot = file.basename()[:-nsuffix]
+        content = file.text()
+        expected = ['.. automodule:: ' + pkgroot]
+        unexpected = []
+        if recursive:
+            lst = expected
+        else:
+            lst = unexpected
+        lst.append(module_rubic)
+        if depth == max_depth:
+            lst.append(last_module_summary.format(pkgroot))
+        else:
+            lst.append(module_summary.format(pkgroot, pkgroot))
+        for text in expected:
+            assert text in content
+        for text in unexpected:
+            assert text not in content
+
+    def assert_module_content(file):
+        modname = file.basename()[:-nsuffix]
+        content = file.text()
+        assert '.. automodule:: '+modname in content
+        assert '   .. rubric:: Classes' in content
+        assert class_summary in content
+
+    root = app.srcdir / 'generated'
+    pkgroot = ''
+    expected_paths = []
+    if app.config.autosummary_recursive:
+        max_depth = 3
+    else:
+        max_depth = 0
+    for depth in range(max_depth+1):
+        if pkgroot:
+            pkgroot = '.'.join((pkgroot, package_name))
+        else:
+            pkgroot = package_name
+        package = root / '.'.join((pkgroot, extension))
+        expected_paths.append(package)
+        assert_package_content(package)
+        if recursive:
+            module = root / module_rubic_name / \
+                '.'.join((pkgroot, module_name, extension))
+            expected_paths.append(root / module_rubic_name)
+            expected_paths.append(module)
+            assert_module_content(module)
+        root /= module_rubic_name
+
+    # Check generated files and directories
+    generated_paths = []
+    root = app.srcdir / 'generated'
+    for dir, subdirs, files in os.walk(str(root)):
+        root = root / dir
+        for name in files:
+            generated_paths.append(root / name)
+        for name in subdirs:
+            generated_paths.append(root / name)
+    assert set(generated_paths) == set(expected_paths)
+
+
+@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive')
+def test_autosummary_recursive(make_app, app_params):
```

Comment:
I feel this testcase is too complicated... How about making it simple?

I think main concern of this testcase is recursive generation. So it is fine to just check existence of stub files.
And it would be great if check root stub file contains `autosummary` definition for recursive call.
```
@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive', confoverrides={'autosummary_recursive': True})
def test_autosummary_recursive_enabled(...):
    app.build()
    assert (app.srcdir / 'generated' / 'package.rst').exists()
    assert (app.srcdir / 'generated' / 'package.module.rst').exists()
    ...

    content = (app.srcdir / 'generated' / 'package.rst').text()
    assert '.. autosummary::\n  :toctree: module\n...' in content


@pytest.mark.sphinx('dummy', testroot='ext-autosummary-recursive', confoverrides={'autosummary_recursive': True})
def test_autosummary_recursive_disabled(...):
    app.build()
    assert (app.srcdir / 'generated' / 'package.rst').exists()
    assert not (app.srcdir / 'generated' / 'package.module.rst').exists()
    ...

```

Reply:
Done